feat(allocator): add `Vec2`

Dunqing opened the pull request:
related: https://github.com/oxc-project/oxc/issues/9706

These two files are copied from Bumplo's `vec` and `raw_vec`, they will be used to replace allocator_api2's `Vec` later PRs.

Dunqing commented:
* **#9647** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9647?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9646** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9646?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9646?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9646).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented on crates/oxc_allocator/src/vec2/mod.rs:
Can you add a link to where you copied the file from?

overlookmotel replied:
Done.

overlookmotel reviewed (approved):
I double-checked these 2 files are copied completely verbatim from `bumpalo`. Only differences are the addition of comments at top of files saying where the code was copied from, and running rustfmt on them.

Merging just this one for now. Would like to look into why lifetime changes are required in #9656 before merging the rest of the stack.

Dunqing committed:
feat(allocator): add `Vec2` (#9646)

related: https://github.com/oxc-project/oxc/issues/9706

These two files are copied from Bumplo's `vec` and `raw_vec`, they will be used to replace allocator_api2's `Vec` later PRs.
